# UCL AI Society Machine Learning Tutorials
### Session 01. Introduction to Numpy, Pandas, Matplotlib

### Contents
1. Numpy
2. Pandas
3. Matplotlib
4. EDA(Exploratory Data Analysis)

### Aim
At the end of this session, you will be able to:
- Understand the basics of numpy.
- Understand the basics of pandas.
- Understand the basics of matplotlib.
- Perform an Exploratory Data Analysis (EDA).


## 2. Pandas
Pandas is another essential open-source library in Python, and today it is widely used by data scientists and ML Engineers. It is built by Wes McKinney based on numpy. The name 'Pandas' is originated from the term "Panel Data", an econometrics term for datasets that include observations over multiple time periods for the same object.

### 2.1 Basics of Pandas

In [ ]:
# run this cell if you haven't installed pandas library
! pip install pandas

In [1]:
import pandas as pd
import numpy as np

In [2]:
print(pd.__version__)

1.1.3


The main data structures of pandas are **Series** and **DataFrame**, where data are stored and manipulated. A `Series` can simply understood as a column and a `DataFrame` as a table that has many Series.

In [3]:
a = pd.Series([1, 2, 3, np.nan, 5, 6])

In [4]:
a

0    1.0
1    2.0
2    3.0
3    NaN
4    5.0
5    6.0
dtype: float64

**Let's check the difference!**

In [5]:
# Creating a Series using pandas.Series()
# This is just one of many ways to initialise a pandas series
module_score_dic = {'Database': 90, 'Security': 70, 'Math': 100, 'Machine Learning': 80}
module_score = pd.Series(module_score_dic)
print("Module_score: \n", module_score, '\n')
print("type: ", type(module_score), '\n')

# Creating a DataFrame using pandas DataFrame()
dataframe = pd.DataFrame(module_score, columns=['score'])
# dataframe = pd.DataFrame(module_score, index=[x for x in module_score.keys()], columns=['score'])
print("dataframe: \n", dataframe, '\n')
print("type: ", type(dataframe))

Module_score: 
 Database             90
Security             70
Math                100
Machine Learning     80
dtype: int64 

type:  <class 'pandas.core.series.Series'> 

dataframe: 
                   score
Database             90
Security             70
Math                100
Machine Learning     80 

type:  <class 'pandas.core.frame.DataFrame'>


Series can also be a Dataframe that has only one attribute.  
**Now let's make a Dataframe that has multiple attributes**

In [6]:
solar_data = {
    'Name' : ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune"],
    'Satellite' : [0, 0, 1, 2, 79, 60, 27, 14],
    'AU' : [0.4, 0.7, 1, 1.5, 5.2, 9.5, 19.2, 30.1],
    'Diameter' : [4.9, 12.1, 12.7, 6.8, 139.8, 116.5, 50.7, 49.2]
}

In [7]:
solar_system = pd.DataFrame(solar_data, index = [i for i in range(1, 9)])
solar_system

Name  Satellite    AU  Diameter
1  Mercury          0   0.4       4.9
2    Venus          0   0.7      12.1
3    Earth          1   1.0      12.7
4     Mars          2   1.5       6.8
5  Jupiter         79   5.2     139.8
6   Saturn         60   9.5     116.5
7   Uranus         27  19.2      50.7
8  Neptune         14  30.1      49.2

In [8]:
solar_system.dtypes # check data type

Name          object
Satellite      int64
AU           float64
Diameter     float64
dtype: object

We can select what to read from the DataFrame by using methods that df.DataFrame has.

- `head()` : Returns the first n data
- `tail()` : Returns the last n data
- `index` : Returns the index
- `columns` : Returns the column
- `loc` : Returns the information of that row
- `values` : Returns only the values without index and column names
- `describe()` : Outputs satatistical summary a DataFrame
- `sort_values(self, by, axis = 0, ascending = True, inplace = False)` : Sort the DataFrame
- `drop()` : Drops selected row

In [9]:
df = solar_system

In [10]:
df.head() # the default value in the bracket is 5

Name  Satellite   AU  Diameter
1  Mercury          0  0.4       4.9
2    Venus          0  0.7      12.1
3    Earth          1  1.0      12.7
4     Mars          2  1.5       6.8
5  Jupiter         79  5.2     139.8

In [11]:
df.tail(2)

Name  Satellite    AU  Diameter
7   Uranus         27  19.2      50.7
8  Neptune         14  30.1      49.2

In [12]:
df.index

Int64Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int64')

In [13]:
df.index[0]

1

In [14]:
df.columns

Index(['Name', 'Satellite', 'AU', 'Diameter'], dtype='object')

In [15]:
# TODO: df.loc[0] gives you an error, but df.iloc[0] works fine. You see why? 
# Try playing around with different numbers and find out why before you google it!
# Google search keyword: df.loc vs df.iloc

df.loc[1]

Name         Mercury
Satellite          0
AU               0.4
Diameter         4.9
Name: 1, dtype: object

In [16]:
df.iloc[0]

Name         Mercury
Satellite          0
AU               0.4
Diameter         4.9
Name: 1, dtype: object

In [17]:
df.values

array([['Mercury', 0, 0.4, 4.9],
       ['Venus', 0, 0.7, 12.1],
       ['Earth', 1, 1.0, 12.7],
       ['Mars', 2, 1.5, 6.8],
       ['Jupiter', 79, 5.2, 139.8],
       ['Saturn', 60, 9.5, 116.5],
       ['Uranus', 27, 19.2, 50.7],
       ['Neptune', 14, 30.1, 49.2]], dtype=object)

In [18]:
df.describe()

Satellite         AU   Diameter
count   8.000000   8.000000    8.00000
mean   22.875000   8.450000   49.08750
std    30.670775  10.853702   52.38417
min     0.000000   0.400000    4.90000
25%     0.750000   0.925000   10.77500
50%     8.000000   3.350000   30.95000
75%    35.250000  11.925000   67.15000
max    79.000000  30.100000  139.80000

In [19]:
df.sort_values(by = 'Diameter', ascending = False)

Name  Satellite    AU  Diameter
5  Jupiter         79   5.2     139.8
6   Saturn         60   9.5     116.5
7   Uranus         27  19.2      50.7
8  Neptune         14  30.1      49.2
3    Earth          1   1.0      12.7
2    Venus          0   0.7      12.1
4     Mars          2   1.5       6.8
1  Mercury          0   0.4       4.9

In [20]:
# TO DO: re-sort the DataFrame by the number of satellite in descending order.
df.sort_values(by = 'Satellite', ascending = False)

Name  Satellite    AU  Diameter
5  Jupiter         79   5.2     139.8
6   Saturn         60   9.5     116.5
7   Uranus         27  19.2      50.7
8  Neptune         14  30.1      49.2
4     Mars          2   1.5       6.8
3    Earth          1   1.0      12.7
1  Mercury          0   0.4       4.9
2    Venus          0   0.7      12.1

Before 2006, Pluto was classified as a planet of the solar system. Let's bring him back to our solar system, by adding Pluto to our DataFrame.

In [21]:
df.loc[9] = ["Pluto", 0, 39.5, 2.38]
df

Name  Satellite    AU  Diameter
1  Mercury          0   0.4      4.90
2    Venus          0   0.7     12.10
3    Earth          1   1.0     12.70
4     Mars          2   1.5      6.80
5  Jupiter         79   5.2    139.80
6   Saturn         60   9.5    116.50
7   Uranus         27  19.2     50.70
8  Neptune         14  30.1     49.20
9    Pluto          0  39.5      2.38

Let's reclassify pluto as a dwarf planet again.

In [22]:
# To drop pluto, you either do df.drop(index=idx) or df.drop(df.index[idx])
df.drop(9)

Name  Satellite    AU  Diameter
1  Mercury          0   0.4       4.9
2    Venus          0   0.7      12.1
3    Earth          1   1.0      12.7
4     Mars          2   1.5       6.8
5  Jupiter         79   5.2     139.8
6   Saturn         60   9.5     116.5
7   Uranus         27  19.2      50.7
8  Neptune         14  30.1      49.2

### 2.2 Read Data with Pandas
Pandas supports loading, reading, and writing data from/to various file format, including CSV, JSON and SQL, by converting it to a DataFrame. 
1. `pd.read_csv()` : Read CSV files
2. `pd.read_json()` : Read JSON files
3. `pd.read_sql_query()` : Read SQL files

In [23]:
# TODO: Try each option, and see what the difference is.
# Option 1
movie = pd.read_csv("./data/IMDB-Movie-Data.csv", index_col = "Title")

#Option 2
# movie = pd.read_csv("./data/IMDB-Movie-Data.csv")
print(type(movie))
movie

<class 'pandas.core.frame.DataFrame'>


Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

In [24]:
# To Do: Extract third row
movie.iloc[2]
# movie.loc["Split"]
# movie.loc[2]  --> Is this going to work? if not, why not?

Rank                                                                  3
Genre                                                   Horror,Thriller
Description           Three girls are kidnapped by a man with a diag...
Director                                             M. Night Shyamalan
Actors                James McAvoy, Anya Taylor-Joy, Haley Lu Richar...
Year                                                               2016
Runtime (Minutes)                                                   117
Rating                                                              7.3
Votes                                                            157606
Revenue (Millions)                                               138.12
Metascore                                                            62
Name: Split, dtype: object

In [25]:
# To Do: Sort the table by Ratings, in descending order
# Do you agree with the rankings? :)
movie.sort_values(by = "Rating", ascending = False)

Rank                     Genre  \
Title                                                     
The Dark Knight            55        Action,Crime,Drama   
Inception                  81   Action,Adventure,Sci-Fi   
Dangal                    118    Action,Biography,Drama   
Interstellar               37    Adventure,Drama,Sci-Fi   
Kimi no na wa              97   Animation,Drama,Fantasy   
...                       ...                       ...   
Wrecker                   969    Action,Horror,Thriller   
Tall Men                  648   Fantasy,Horror,Thriller   
Dragonball Evolution      872  Action,Adventure,Fantasy   
Don't Fuck in the Woods    43                    Horror   
Disaster Movie            830                    Comedy   

                                                               Description  \
Title                                                                        
The Dark Knight          When the menace known as the Joker wreaks havo...   
Inception                A thief, who steals corporate secrets through ...   
Dangal                   Former wrestler Mahavir Singh Phogat and his t...   
Interstellar             A team of explorers travel through a wormhole ...   
Kimi no na wa            Two strangers find themselves linked in a biza...   
...                                                                    ...   
Wrecker                  Best friends Emily and Lesley go on a road tri...   
Tall Men                 A challenged man is stalked by tall phantoms i...   
Dragonball Evolution     The young warrior Son Goku sets out on a quest...   
Don't Fuck in the Woods  A group of friends are going on a camping trip...   
Disaster Movie           Over the course of one evening, an unsuspectin...   

                                  Director  \
Title                                        
The Dark Knight          Christopher Nolan   
Inception                Christopher Nolan   
Dangal                       Nitesh Tiwari   
Interstellar             Christopher Nolan   
Kimi no na wa               Makoto Shinkai   
...                                    ...   
Wrecker                     Micheal Bafaro   
Tall Men                 Jonathan Holbrook   
Dragonball Evolution            James Wong   
Don't Fuck in the Woods      Shawn Burkett   
Disaster Movie             Jason Friedberg   

                                                                    Actors  \
Title                                                                        
The Dark Knight          Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
Inception                Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
Dangal                   Aamir Khan, Sakshi Tanwar, Fatima Sana Shaikh,...   
Interstellar             Matthew McConaughey, Anne Hathaway, Jessica Ch...   
Kimi no na wa            Ryûnosuke Kamiki, Mone Kamishiraishi, Ryô Nari...   
...                                                                    ...   
Wrecker                  Anna Hutchison, Andrea Whitburn, Jennifer Koen...   
Tall Men                 Dan Crisafulli, Kay Whitney, Richard Garcia, P...   
Dragonball Evolution     Justin Chatwin, James Marsters, Yun-Fat Chow, ...   
Don't Fuck in the Woods  Brittany Blanton, Ayse Howard, Roman Jossart,N...   
Disaster Movie           Carmen Electra, Vanessa Lachey,Nicole Parker, ...   

                         Year  Runtime (Minutes)  Rating    Votes  \
Title                                                               
The Dark Knight          2008                152     9.0  1791916   
Inception                2010                148     8.8  1583625   
Dangal                   2016                161     8.8    48969   
Interstellar             2014                169     8.6  1047747   
Kimi no na wa            2016                106     8.6    34110   
...                       ...                ...     ...      ...   
Wrecker                  2015                 83     3.5     1210   
Tall Me

In [26]:
# To Do: Sort the table by 'Revenue(Millions)', in ascending order and print the first 3 rows out
movie.sort_values(by = "Revenue (Millions)", ascending = True).head(3)

Rank                  Genre  \
Title                                           
A Kind of Murder   232   Crime,Drama,Thriller   
Into the Forest    962  Drama,Sci-Fi,Thriller   
Love, Rosie        678         Comedy,Romance   

                                                        Description  \
Title                                                                 
A Kind of Murder  In 1960s New York, Walter Stackhouse is a succ...   
Into the Forest   After a massive power outage, two sisters lear...   
Love, Rosie       Rosie and Alex have been best friends since th...   

                          Director  \
Title                                
A Kind of Murder      Andy Goddard   
Into the Forest    Patricia Rozema   
Love, Rosie       Christian Ditter   

                                                             Actors  Year  \
Title                                                                       
A Kind of Murder  Patrick Wilson, Jessica Biel, Haley Bennett, V...  2016   
Into the Forest   Ellen Page, Evan Rachel Wood, Max Minghella,Ca...  2015   
Love, Rosie       Lily Collins, Sam Claflin, Christian Cooke, Ja...  2014   

                  Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                    
A Kind of Murder                 95     5.2   3305                0.00   
Into the Forest                 101     5.9  10220                0.01   
Love, Rosie                     102     7.2  80415                0.01   

                  Metascore  
Title                        
A Kind of Murder       50.0  
Into the Forest        59.0  
Love, Rosie            44.0

In [27]:
# The value_counts() function is used to get a Series containing counts of unique values
movie['Genre'].value_counts().head()

Action,Adventure,Sci-Fi    50
Drama                      48
Comedy,Drama,Romance       35
Comedy                     32
Drama,Romance              31
Name: Genre, dtype: int64

In [28]:
# This is called a "Masking Operation"
# filter out movies that have runtime under 170 minutes and sort the result by rating in descending order.
movie[movie['Runtime (Minutes)'] >= 170].sort_values(by="Rating", ascending=False)

Rank                   Genre  \
Title                                                   
3 Idiots                  431            Comedy,Drama   
The Wolf of Wall Street    83  Biography,Comedy,Crime   
The Hateful Eight          89     Crime,Drama,Mystery   
La vie d'Adèle            312           Drama,Romance   
Grindhouse                829  Action,Horror,Thriller   
Cloud Atlas               268            Drama,Sci-Fi   
Inland Empire             966  Drama,Mystery,Thriller   

                                                               Description  \
Title                                                                        
3 Idiots                 Two friends are searching for their long lost ...   
The Wolf of Wall Street  Based on the true story of Jordan Belfort, fro...   
The Hateful Eight        In the dead of a Wyoming winter, a bounty hunt...   
La vie d'Adèle           Adèle's life is changed when she meets Emma, a...   
Grindhouse               Quentin Tarantino and Robert Rodriguez's homag...   
Cloud Atlas              An exploration of how the actions of individua...   
Inland Empire            As an actress starts to adopt the persona of h...   

                                    Director  \
Title                                          
3 Idiots                     Rajkumar Hirani   
The Wolf of Wall Street      Martin Scorsese   
The Hateful Eight          Quentin Tarantino   
La vie d'Adèle           Abdellatif Kechiche   
Grindhouse                  Robert Rodriguez   
Cloud Atlas                       Tom Tykwer   
Inland Empire                    David Lynch   

                                                                    Actors  \
Title                                                                        
3 Idiots                   Aamir Khan, Madhavan, Mona Singh, Sharman Joshi   
The Wolf of Wall Street  Leonardo DiCaprio, Jonah Hill, Margot Robbie,M...   
The Hateful Eight        Samuel L. Jackson, Kurt Russell, Jennifer Jaso...   
La vie d'Adèle           Léa Seydoux, Adèle Exarchopoulos, Salim Kechio...   
Grindhouse               Kurt Russell, Rose McGowan, Danny Trejo, Zoë Bell   
Cloud Atlas               Tom Hanks, Halle Berry, Hugh Grant, Hugo Weaving   
Inland Empire            Laura Dern, Jeremy Irons, Justin Theroux, Karo...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
3 Idiots                 2009                170     8.4  238789   
The Wolf of Wall Street  2013                180     8.2  865134   
The Hateful Eight        2015                187     7.8  341170   
La vie d'Adèle           2013                180     7.8  103150   
Grindhouse               2007                191     7.6  160350   
Cloud Atlas              2012                172     7.5  298651   
Inland Empire            2006                180     7.0   44227   

                         Revenue (Millions)  Metascore  
Title                                                   
3 Idiots                               6.52       67.0  
The Wolf of Wall Street              116.87       75.0  
The Hateful Eight                     54.12       68.0  
La vie d'Adèle                         2.20       88.0  
Grindhouse                            25.03        NaN  
Cloud Atlas                           27.10       55.0  
Inland Empire                           NaN        NaN

In [29]:
# To Do: Extract the movies whose 'Metascore' is bigger than 95, and sort the result from the most recent to the least recent
movie[movie['Metascore'] >= 95].sort_values(by="Year", ascending=False)

Rank                    Genre  \
Title                                                  
Manchester by the Sea    22                    Drama   
Moonlight                42                    Drama   
Carol                   502            Drama,Romance   
Boyhood                 657                    Drama   
12 Years a Slave        112  Biography,Drama,History   
Gravity                 510    Drama,Sci-Fi,Thriller   
Zero Dark Thirty        407   Drama,History,Thriller   
The Social Network      325          Biography,Drama   
Ratatouille             490  Animation,Comedy,Family   
Pan's Labyrinth         231        Drama,Fantasy,War   

                                                             Description  \
Title                                                                      
Manchester by the Sea  A depressed uncle is asked to take care of his...   
Moonlight              A chronicle of the childhood, adolescence and ...   
Carol                  An aspiring photographer develops an intimate ...   
Boyhood                The life of Mason, from early childhood to his...   
12 Years a Slave       In the antebellum United States, Solomon North...   
Gravity                Two astronauts work together to survive after ...   
Zero Dark Thirty       A chronicle of the decade-long hunt for al-Qae...   
The Social Network     Harvard student Mark Zuckerberg creates the so...   
Ratatouille            A rat who can cook makes an unusual alliance w...   
Pan's Labyrinth        In the falangist Spain of 1944, the bookish yo...   

                                 Director  \
Title                                       
Manchester by the Sea    Kenneth Lonergan   
Moonlight                   Barry Jenkins   
Carol                         Todd Haynes   
Boyhood                 Richard Linklater   
12 Years a Slave            Steve McQueen   
Gravity                    Alfonso Cuarón   
Zero Dark Thirty          Kathryn Bigelow   
The Social Network          David Fincher   
Ratatouille                     Brad Bird   
Pan's Labyrinth        Guillermo del Toro   

                                                                  Actors  \
Title                                                                      
Manchester by the Sea  Casey Affleck, Michelle Williams, Kyle Chandle...   
Moonlight              Mahershala Ali, Shariff Earp, Duan Sanderson, ...   
Carol                  Cate Blanchett, Rooney Mara, Sarah Paulson, Ky...   
Boyhood                Ellar Coltrane, Patricia Arquette, Ethan Hawke...   
12 Years a Slave       Chiwetel Ejiofor, Michael Kenneth Williams, Mi...   
Gravity                Sandra Bullock, George Clooney, Ed Harris, Ort...   
Zero Dark Thirty       Jessica Chastain, Joel Edgerton, Chris Pratt, ...   
The Social Network     Jesse Eisenberg, Andrew Garfield, Justin Timbe...   
Ratatouille             Brad Garrett, Lou Romano, Patton Oswalt,Ian Holm   
Pan's Labyrinth        Ivana Baquero, Ariadna Gil, Sergi López,Maribe...   

                       Year  Runtime (Minutes)  Rating   Votes  \
Title                                                            
Manchester by the Sea  2016                137     7.9  134213   
Moonlight              2016                111     7.5  135095   
Carol                  2015                118     7.2   77995   
Boyhood                2014                165     7.9  286722   
12 Years a Slave       2013                134     8.1  486338   
Gravity                2013                 91     7.8  622089   
Zero Dark Thirty       2012                157     7.4  226661   
The Social Network     2010                120     7.7  510100   
Ratatouille            2007                111     8.0  504039   
Pan's Labyrinth        2006                118     8.2  498879   

                       Revenue (Millions)  Metascore  
Title                                                 
Manchester by the Sea               47.70       96.0  
Moonlight                

In [30]:
# To Do: Extract movies whose directed by one of UCL Alumni ---> Hint: Tenet, Inception
movie[(movie['Director'] == 'Christopher Nolan')]

Rank                    Genre  \
Title                                                  
Interstellar             37   Adventure,Drama,Sci-Fi   
The Dark Knight          55       Action,Crime,Drama   
The Prestige             65     Drama,Mystery,Sci-Fi   
Inception                81  Action,Adventure,Sci-Fi   
The Dark Knight Rises   125          Action,Thriller   

                                                             Description  \
Title                                                                      
Interstellar           A team of explorers travel through a wormhole ...   
The Dark Knight        When the menace known as the Joker wreaks havo...   
The Prestige           Two stage magicians engage in competitive one-...   
Inception              A thief, who steals corporate secrets through ...   
The Dark Knight Rises  Eight years after the Joker's reign of anarchy...   

                                Director  \
Title                                      
Interstellar           Christopher Nolan   
The Dark Knight        Christopher Nolan   
The Prestige           Christopher Nolan   
Inception              Christopher Nolan   
The Dark Knight Rises  Christopher Nolan   

                                                                  Actors  \
Title                                                                      
Interstellar           Matthew McConaughey, Anne Hathaway, Jessica Ch...   
The Dark Knight        Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
The Prestige           Christian Bale, Hugh Jackman, Scarlett Johanss...   
Inception              Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
The Dark Knight Rises  Christian Bale, Tom Hardy, Anne Hathaway,Gary ...   

                       Year  Runtime (Minutes)  Rating    Votes  \
Title                                                             
Interstellar           2014                169     8.6  1047747   
The Dark Knight        2008                152     9.0  1791916   
The Prestige           2006                130     8.5   913152   
Inception              2010                148     8.8  1583625   
The Dark Knight Rises  2012                164     8.5  1222645   

                       Revenue (Millions)  Metascore  
Title                                                 
Interstellar                       187.99       74.0  
The Dark Knight                    533.32       82.0  
The Prestige                        53.08       66.0  
Inception                          292.57       74.0  
The Dark Knight Rises              448.13       78.0

#### 2.2.1 Pandas Exercise

To Do: Extract the movie list that meets requirements below:
- 1. Released after 2010 (key = 'Year') (including year 2010)
- 2. Runtime is shorter than 150 minutes (key = 'Runtime (Minutes)')
- 3. Rating is above 8.0 (key = 'Rating')  

Print out only the first 3 movies from the result.

In [31]:
movie[
    (movie['Year'] >= 2010) & 
    (movie['Runtime (Minutes)'] <= 150) & 
    (movie['Rating'] >= 8.0)
][:3]

Rank                    Genre  \
Title                                                    
Guardians of the Galaxy     1  Action,Adventure,Sci-Fi   
La La Land                  7       Comedy,Drama,Music   
Hacksaw Ridge              17  Biography,Drama,History   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
La La Land               A jazz pianist falls for an aspiring actress i...   
Hacksaw Ridge            WWII American Army Medic Desmond T. Doss, who ...   

                                Director  \
Title                                      
Guardians of the Galaxy       James Gunn   
La La Land               Damien Chazelle   
Hacksaw Ridge                 Mel Gibson   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
La La Land               Ryan Gosling, Emma Stone, Rosemarie DeWitt, J....   
Hacksaw Ridge            Andrew Garfield, Sam Worthington, Luke Bracey,...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
La La Land               2016                128     8.3  258682   
Hacksaw Ridge            2016                139     8.2  211760   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
La La Land                           151.06       93.0  
Hacksaw Ridge                         67.12       71.0

### 2.3 How to deal with Missing Data
To represent missing data, pandas use np.nan. Data scientists and machine learning engineers sometimes just remove missing data. However, it heavily depends on which data are missing, how big the missing data are and so on. You can fill the missing part with 0, with the mean value of the column or with mean value of only 10 nearest value in the column. It is important for you to choose the way how you are going to deal with missing data.
- `isnull()`: returns True or False, depending on the cell's null status. 
- `sum()`: This can be used as a trick when you count the number of True's. Once the Dataframe is filtered through isnull() function, sum of all True's in a column gives you how many fields have missing data in them.
- `dropna()`: deletes any row that contains any single null value.
- `fillna(value)`: Fill missing value with the given values.

In [32]:
movie.isnull()

Rank  Genre  Description  Director  Actors   Year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                           
Guardians of the Galaxy              False   False  False               False   
Prometheus                           False   False  False               False   
Split                                False   False  False               False   
Sing                                 False   False  False               False   
Suicide Squad                        False   False  False               False   
...                                    ...     ...    ...                 ...   
Secret in Their Eyes                 False   False  False                True   
Hostel: Part II                      False   False  False               False   
Step Up 2: The Streets               False   False  False               False   
Search Party                         False   False  False                True   
Nine Lives                           False   False  False               False   

                         Metascore  
Title                               
Guardians of the Galaxy      False  
Prometheus                   False  
Split                        False  
Sing                         False  
Suicide Squad                False  
...                            ...  
Secret in Their Eyes         False  
Hostel: Part II              False  
Step Up 2: The Streets       False  
Search Party                 False  
Nine Lives                   False  

[1000 rows x 11 columns]

In [33]:
movie.isnull().sum()

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

In [34]:
movie.shape

(1000, 11)

In [35]:
# Take a look at "Take Me Home Tonight" and "Search Party"
movie.fillna(value = 0)

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

In [36]:
movie.dropna(inplace = True)

In [37]:
movie.shape

(838, 11)

After dropping the rows that contain missing data, the shape of the dataFrame has changed, from (1000, 11) to (838, 11)

### 2.4 Merging Data
Some of you who know SQL might have felt that pandas is quite similar to query language.
What is the most popular thing that you do in most of the relational database query language?  
Yes! (terminologies alert!) Inner JOIN, Outer JOIN, Left JOIN, Right JOIN, Full JOIN...
- `concat()` : Concatenation. Used to merge two or more pandas object.
- `merge()` : Behaves very simlar to SQL.

We'll gonna create random dataframe, named df1 and df2

In [38]:
df1 = pd.DataFrame(np.random.randn(10, 2))
df1

0         1
0  0.123106 -0.736940
1 -1.052358  0.279380
2  1.911297  0.793828
3  0.817797  0.283461
4  0.871031 -0.628277
5 -0.057334  1.357887
6  0.851312  1.493082
7 -0.052685  0.764155
8 -1.768542  0.678152
9 -0.594167  0.758330

In [39]:
df2 = pd.DataFrame(np.random.randn(10, 3))
df2

0         1         2
0 -0.647700  0.390959  1.180534
1 -0.987608 -1.836777 -0.131682
2 -0.018339 -0.884299 -2.493277
3  0.149265  0.633611 -0.112990
4  1.416782  1.190578  0.593015
5  0.593569 -0.882884 -0.328260
6  1.118777 -0.880869 -1.295504
7 -0.410248 -0.165534  0.479961
8  0.257958 -0.237406 -1.783602
9  2.293270  0.137492 -0.639376

In [40]:
pd.concat([df1, df2])

0         1         2
0  0.123106 -0.736940       NaN
1 -1.052358  0.279380       NaN
2  1.911297  0.793828       NaN
3  0.817797  0.283461       NaN
4  0.871031 -0.628277       NaN
5 -0.057334  1.357887       NaN
6  0.851312  1.493082       NaN
7 -0.052685  0.764155       NaN
8 -1.768542  0.678152       NaN
9 -0.594167  0.758330       NaN
0 -0.647700  0.390959  1.180534
1 -0.987608 -1.836777 -0.131682
2 -0.018339 -0.884299 -2.493277
3  0.149265  0.633611 -0.112990
4  1.416782  1.190578  0.593015
5  0.593569 -0.882884 -0.328260
6  1.118777 -0.880869 -1.295504
7 -0.410248 -0.165534  0.479961
8  0.257958 -0.237406 -1.783602
9  2.293270  0.137492 -0.639376

In [41]:
pd.concat([df1, df2], axis = 1)   # axis setting is very common in pandas

0         1         0         1         2
0  0.123106 -0.736940 -0.647700  0.390959  1.180534
1 -1.052358  0.279380 -0.987608 -1.836777 -0.131682
2  1.911297  0.793828 -0.018339 -0.884299 -2.493277
3  0.817797  0.283461  0.149265  0.633611 -0.112990
4  0.871031 -0.628277  1.416782  1.190578  0.593015
5 -0.057334  1.357887  0.593569 -0.882884 -0.328260
6  0.851312  1.493082  1.118777 -0.880869 -1.295504
7 -0.052685  0.764155 -0.410248 -0.165534  0.479961
8 -1.768542  0.678152  0.257958 -0.237406 -1.783602
9 -0.594167  0.758330  2.293270  0.137492 -0.639376

In [42]:
demis = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Demis' : [75, 97, 64, 81]}
)
demis

Modules  Demis
0   Bioinformatics     75
1  Robotic Systems     97
2         Security     64
3        Compilers     81

In [43]:
sedol = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Sedol' : [63, 78, 84, 95]})
sedol

Modules  Sedol
0   Bioinformatics     63
1  Robotic Systems     78
2         Security     84
3        Compilers     95

In [44]:
pd.merge(demis, sedol, on = 'Modules')

Modules  Demis  Sedol
0   Bioinformatics     75     63
1  Robotic Systems     97     78
2         Security     64     84
3        Compilers     81     95

In [132]:
# To Do: Define your own dataframe and use functions introduced above to concatenate or merge them.

In [ ]:
# Your own trial code here!

### What to do next?
Below websites would be helpful for your further study on pandas library:
- [Pandas official website](https://pandas.pydata.org)
- [10 minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html)
- [Data Wrangling with Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)